core/src/main/java/org/apache/iceberg/schema/UnionByNameVisitor.java

```diff
@@ -163,8 +164,7 @@ private void updateColumn(Types.NestedField field, Types.NestedField existingFie
     String fullName = partnerSchema.findColumnName(existingField.fieldId());
 
     boolean needsOptionalUpdate = field.isOptional() && existingField.isRequired();
-    boolean needsTypeUpdate =
-        field.type().isPrimitiveType() && !field.type().equals(existingField.type());
+    boolean needsTypeUpdate = !compatibleType(field.type(), existingField.type());
```

Comment:
Do we also need the `field.type().isPrimitiveType()` here?
```
boolean needsTypeUpdate = field.type().isPrimitiveType() && !compatibleType(field.type(), existingField.type());
```

Reply:
Good catch, we included this in the latest update, it's included in the helper method.